src/exec/use_pty/pipe.rs

```diff
@@ -219,16 +271,169 @@ impl<R: Read, W: Write> Buffer<R, W> {
 
     /// Flush this buffer, ensuring that all the contents of its internal buffer are written.
     fn flush(&mut self, write: &mut W) -> io::Result<()> {
-        // This is the busy section of the buffer.
-        let buffer = &self.internal[self.start..self.end];
+        // Remove bytes from the buffer and write them.
+        self.internal.remove(write)?;
 
-        // Write the complete busy section to `write`.
-        write.write_all(buffer)?;
+        write.flush()
+    }
+}
 
-        // If we were able to write all the busy section, we can mark the whole buffer as free.
-        self.start = 0;
-        self.end = 0;
+#[cfg(test)]
+mod tests {
+    use super::RingBuffer;
 
-        write.flush()
+    #[test]
+    fn empty_buffer_is_empty() {
+        let buf = RingBuffer::new();
+
+        assert!(buf.is_empty());
+    }
+
+    #[test]
+    fn full_buffer_is_full() {
+        let mut buf = RingBuffer::new();
+
+        let inserted_len = buf.insert(&mut [0x45; RingBuffer::LEN].as_slice()).unwrap();
+        assert_eq!(inserted_len, RingBuffer::LEN);
+
+        assert!(buf.is_full());
+    }
+
+    #[test]
+    fn buffer_is_fifo() {
+        let mut buf = RingBuffer::new();
+
+        let expected = (0..=u8::MAX).collect::<Vec<u8>>();
+        let inserted_len = buf.insert(&mut expected.as_slice()).unwrap();
+        assert_eq!(inserted_len, expected.len());
+
+        let mut found = vec![];
+        let removed_len = buf.remove(&mut found).unwrap();
+        assert_eq!(removed_len, expected.len());
+
+        assert_eq!(expected, found);
+    }
+
+    #[test]
+    fn insert_into_empty_buffer_with_offset() {
+        const HALF_LEN: usize = RingBuffer::LEN / 2;
+        let mut buf = RingBuffer::new();
+
+        // This should leave the buffer empty but with the start field pointing to the middle of
+        // the buffer.
+        // ┌───────────────────┐
+        // │                   │
+        // └───────────────────┘
+        //           ▲
+        //           │
+        //         start
+        buf.insert(&mut [0u8; HALF_LEN].as_slice()).unwrap();
+        buf.remove(&mut vec![]).unwrap();
+
+        // Then we fill the first half of the buffer with ones and the second one with twos in a
+        // single insertion. This tests case 1.1.
```

Comment:
[`cov-mark`](https://crates.io/crates/cov-mark) is nice for testing that this method call hit a certain branch inside its implementation.

Reply:
also this is not quite "testing" case 1.1. you'll hit the branch 1.1 but there are no assertions about what is expected to happen there so I think the values of `start` and `len` as well as the contents of `storage` should be checked after `insert` is called.

I can see the upside to only using public API in the unit tests but currently this unit test relies on both 1.1 and 2.1 working correctly so without a mid-way check, in principle, it could pass even though cases 1.1 and 2.1 are incorrectly implemented.

Reply:
(P.S. I like the ASCII diagrams :100: )